### ISSUE TITLE
cute auxiliary loss idea

### ISSUE THREAD

**Issue by enolan** [enolan/txt2img-unsupervised#1]
the current cap conditioned model is somewhat data inefficient because it doesn't actually see the CLIP embeddings of the training examples, only caps that contain them. we can "fix" this without changing how inference works by appending a final token that encodes the CLIP embedding of the image, and using the cosine distance between output and actual as an auxiliary loss. at inference time we either ignore it or simply don't generate it. 
idk if this would improve quality/prompt following, probably depends on how much overlap there is between the image prediction and CLIP prediction tasks. strikes me that this is pretty much training our model to be a weird version of CLIP at this point...

**Comment by enolan** [comment]
clever variant: choose an integer n that divides the number of image tokens. add n special tokens, evenly spaced within the image tokens, starting directly after the cap token and before the first image token. loss is the same as the first version: cosine distance to the image's CLIP embedding. this scheme forces the model to build up an estimate of the image's embedding over time, conditioning on more image tokens as it goes on. it'll naturally learn to use that estimate to guide the logits. the first special token sees no image tokens, so you might think it'd just learn to predict the cap's center, but the density of the training distribution's embeddings isn't uniform over the unit sphere. so the center of the distribution of potential embeddings for the image is different from the cap's center.

**Comment by enolan** [comment]
two-stage design is much better nvm
